lib/fluent/plugin/filter_kubernetes_metadata.rb

```diff
@@ -253,94 +251,95 @@ def log.trace?
 
     end
 
-    def get_metadata_for_record(match_data, cache_key, create_time, batch_miss_cache)
-      namespace_name = match_data['namespace']
-      pod_name = match_data['pod_name']
+    def get_metadata_for_record(namespace_name, pod_name, container_name, container_id, create_time, batch_miss_cache)
       metadata = {
-        'container_name' => match_data['container_name'],
-        'namespace_name' => namespace_name,
-        'pod_name'       => pod_name
+        'docker' => {'container_id' => container_id},
+        'kubernetes' => {
+          'container_name'  => container_name,
+          'namespace_name'  => namespace_name,
+          'pod_name'        => pod_name
+        }
       }
       if @kubernetes_url.present?
-        pod_metadata = get_pod_metadata(cache_key, namespace_name, pod_name, create_time, batch_miss_cache)
-        metadata.merge!(pod_metadata) if pod_metadata
-      end
-      metadata
-    end
+        pod_metadata = get_pod_metadata(container_id, namespace_name, pod_name, create_time, batch_miss_cache)
 
-    def create_time_from_record(record)
-        time = if @use_journal
-           record['_SOURCE_REALTIME_TIMESTAMP'].nil? ? record['_SOURCE_REALTIME_TIMESTAMP'] : record['__REALTIME_TIMESTAMP']
-        else
-          record['time']
+        if (pod_metadata.include? 'containers') && (pod_metadata['containers'].include? container_id)
+          metadata['kubernetes']['container_image'] = pod_metadata['containers'][container_id]['image']
+          metadata['kubernetes']['container_image_id'] = pod_metadata['containers'][container_id]['image_id']
         end
-        (time.nil? || time.chop.empty?) ? Time.now : Time.parse(time)
-    end
 
-    def filter_stream(tag, es)
-      es
+        metadata['kubernetes'].merge!(pod_metadata) if pod_metadata
+        metadata['kubernetes'].delete('containers')
+      end
+      metadata
     end
 
-    def filter_stream_from_files(tag, es)
-      return es if (es.respond_to?(:empty?) && es.empty?) || !es.is_a?(Fluent::EventStream)
-      new_es = MultiEventStream.new
-
-      match_data = tag.match(@tag_to_kubernetes_name_regexp_compiled)
-      batch_miss_cache = {}
-      if match_data
-        container_id = match_data['docker_id']
-        metadata = {
-          'docker' => {
-            'container_id' => container_id
-          },
-          'kubernetes' => get_metadata_for_record(match_data, container_id, create_time_from_record(es.first[1]), batch_miss_cache)
-        }
+    def create_time_from_record(record, internal_time)
+      time_key = @time_fields.detect{ |ii| record.has_key?(ii) }
+      time = record[time_key]
+      if time.nil? || time.chop.empty?
+        return internal_time
       end
-
-      es.each do |time, record|
-        record = record.merge(Marshal.load(Marshal.dump(metadata))) if metadata
-        new_es.add(time, record)
+      if time_key == '_SOURCE_REALTIME_TIMESTAMP' || time_key == '__REALTIME_TIMESTAMP'
```

Comment:
`if ['_SOURCE_REALTIME_TIMESTAMP','__REALTIME_TIMESTAMP']`.include?(time_key)`

and that array could be reused in the other code bit that pushes them onto an array:  `source.concat(other)`